### PR TITLE
Implement noReturnCheck

### PR DESCRIPTION
**Pull request by doehyunbaek**
I ported the [JSTAR's noReturnCheck](https://github.com/kaist-plrg/jstar/blob/1282919127ea18a7e40c7a55e63a1ddaaf7d9db4/src/main/scala/kr/ac/kaist/jstar/analyzer/AbsSemantics.scala#L66-L71) rule to ESMeta with the appropriate modifications.

Also I added the parsing rule for `sequence of Unicode code points` in the meantime to reduce the number of false alarms. Previously interprocedural analysis into `OrdinaryFunctionCreate` AO made many other AOs that call into this function also evaluates to bottom and this parsing change stops them from happening.

After the two commits, with respect to es2022, this alarm emits total of 152 alarms, of which

<details>
<summary>95 of them contain expressions not yet understandable by ESMeta
</summary>

[NoReturn] AgentSignifier[1069]:RETURN
[NoReturn] Alternative[0,0].CompileSubpattern[2266]:RETURN
[NoReturn] Alternative[1,0].CompileSubpattern[2267]:RETURN
[NoReturn] Assertion[0,0].CompileAssertion[2272]:RETURN
[NoReturn] Assertion[1,0].CompileAssertion[2273]:RETURN
[NoReturn] Assertion[2,0].CompileAssertion[2274]:RETURN
[NoReturn] Assertion[3,0].CompileAssertion[2275]:RETURN
[NoReturn] Assertion[4,0].CompileAssertion[2276]:RETURN
[NoReturn] Assertion[5,0].CompileAssertion[2277]:RETURN
[NoReturn] Assertion[6,0].CompileAssertion[2278]:RETURN
[NoReturn] Assertion[7,0].CompileAssertion[2279]:RETURN
[NoReturn] AsyncBlockStart:cont0[2664]:RETURN
[NoReturn] AsyncConciseBody[1,0].AsyncConciseBodyContainsUseStrict[1738]:RETURN
[NoReturn] AsyncGeneratorStart:cont0[2647]:RETURN
[NoReturn] Atom[0,0].CompileAtom[2289]:RETURN
[NoReturn] Atom[1,0].CompileAtom[2290]:RETURN
[NoReturn] Atom[4,0].CompileAtom[2292]:RETURN
[NoReturn] AtomEscape[2,0].CompileAtom[2295]:RETURN
[NoReturn] AtomEscape[3,0].CompileAtom[2297]:RETURN
[NoReturn] BackreferenceMatcher[2299]:RETURN
[NoReturn] CharacterClassEscape[0,0].CompileToCharSet[2313]:RETURN
[NoReturn] CharacterClassEscape[1,0].CompileToCharSet[2314]:RETURN
[NoReturn] CharacterClassEscape[2,0].CompileToCharSet[2315]:RETURN
[NoReturn] CharacterClassEscape[3,0].CompileToCharSet[2316]:RETURN
[NoReturn] CharacterClassEscape[5,0].CompileToCharSet[2318]:RETURN
[NoReturn] CharacterClassEscape[6,0].CompileToCharSet[2319]:RETURN
[NoReturn] CharacterClassEscape[7,0].CompileToCharSet[2320]:RETURN
[NoReturn] CharacterEscape[0,0].CharacterValue[2238]:RETURN
[NoReturn] CharacterEscape[1,0].CharacterValue[2239]:RETURN
[NoReturn] CharacterEscape[2,0].CharacterValue[2240]:RETURN
[NoReturn] CharacterEscape[5,0].CharacterValue[2248]:RETURN
[NoReturn] CharacterSetMatcher[2298]:RETURN
[NoReturn] ClassAtom[0,0].CharacterValue[2234]:RETURN
[NoReturn] ClassAtom[0,0].CompileToCharSet[2308]:RETURN
[NoReturn] ClassAtomNoDash[0,0].CharacterValue[2235]:RETURN
[NoReturn] ClassAtomNoDash[0,0].CompileToCharSet[2309]:RETURN
[NoReturn] ClassEscape[0,0].CharacterValue[2236]:RETURN
[NoReturn] ClassEscape[0,0].CompileToCharSet[2310]:RETURN
[NoReturn] ClassEscape[1,0].CharacterValue[2237]:RETURN
[NoReturn] ClassEscape[1,0].CompileToCharSet[2311]:RETURN
[NoReturn] ClassEscape[3,0].CompileToCharSet[2312]:RETURN
[NoReturn] ClassRanges[0,0].CompileToCharSet[2303]:RETURN
[NoReturn] ClassTail[0,0].ClassDefinitionEvaluation[1715]:RETURN
[NoReturn] CodePointAt[1193]:RETURN
[NoReturn] ConciseBody[1,0].ConciseBodyContainsUseStrict[1601]:RETURN
[NoReturn] CreateByteDataBlock[104]:RETURN
[NoReturn] CreateMappedArgumentsObject[1146]:RETURN
[NoReturn] CreateSharedByteDataBlock[105]:RETURN
[NoReturn] DebuggerStatement[0,0].Evaluation[1534]:RETURN
[NoReturn] DecimalEscape[0,1].CapturingGroupNumber[2227]:RETURN
[NoReturn] Disjunction[1,0].CompileSubpattern[2265]:RETURN
[NoReturn] FunctionBody[0,0].FunctionBodyContainsUseStrict[1589]:RETURN
[NoReturn] GeneratorStart:cont0[2632]:RETURN
[NoReturn] IdentifierName[1,0].IdentifierCodePoints[1196]:RETURN
[NoReturn] IdentifierPart[0,0].IdentifierCodePoint[1198]:RETURN
[NoReturn] IdentifierStart[0,0].IdentifierCodePoint[1197]:RETURN
[NoReturn] INTRINSICS.Atomics.add[2554]:RETURN
[NoReturn] INTRINSICS.Atomics.sub[2562]:RETURN
[NoReturn] INTRINSICS.Number.prototype.toLocaleString[26]:RETURN
[NoReturn] INTRINSICS.String.prototype.toLocaleLowerCase[32]:RETURN
[NoReturn] INTRINSICS.String.prototype.toLocaleUpperCase[9]:RETURN
[NoReturn] INTRINSICS.String.prototype.toUpperCase[25]:RETURN
[NoReturn] IsValidRegularExpressionLiteral[1262]:RETURN
[NoReturn] LocalTime[2132]:RETURN
[NoReturn] ModuleEnvironmentRecord.CreateImportBinding[1047]:RETURN
[NoReturn] NonemptyClassRanges[1,0].CompileToCharSet[2304]:RETURN
[NoReturn] NonemptyClassRanges[2,0].CompileToCharSet[2305]:RETURN
[NoReturn] NonemptyClassRangesNoDash[1,0].CompileToCharSet[2306]:RETURN
[NoReturn] NonemptyClassRangesNoDash[2,0].CompileToCharSet[2307]:RETURN
[NoReturn] ParsePattern[2329]:RETURN
[NoReturn] PropertyName[1,0].ComputedPropertyContains[853]:RETURN
[NoReturn] RegExpIdentifierName[1,0].RegExpIdentifierCodePoints[2256]:RETURN
[NoReturn] RegExpIdentifierPart[0,0].RegExpIdentifierCodePoint[2258]:RETURN
[NoReturn] RegExpIdentifierPart[1,0].RegExpIdentifierCodePoint[2260]:RETURN
[NoReturn] RegExpIdentifierPart[2,0].RegExpIdentifierCodePoint[2262]:RETURN
[NoReturn] RegExpIdentifierStart[0,0].RegExpIdentifierCodePoint[2257]:RETURN
[NoReturn] RegExpIdentifierStart[1,0].RegExpIdentifierCodePoint[2259]:RETURN
[NoReturn] RegExpIdentifierStart[2,0].RegExpIdentifierCodePoint[2261]:RETURN
[NoReturn] RegularExpressionLiteral[0,0].BodyText[1209]:RETURN
[NoReturn] RegularExpressionLiteral[0,0].FlagText[1210]:RETURN
[NoReturn] SerializeJSONArray[2573]:RETURN
[NoReturn] SetFunctionName[1123]:RETURN
[NoReturn] Term[2,0].CompileSubpattern[2270]:RETURN
[NoReturn] TypedArrayElementSize[2444]:RETURN
[NoReturn] TypedArrayElementType[2445]:RETURN
[NoReturn] UnicodeEscapeSequence[0,0].IdentifierCodePoint[1199]:RETURN
[NoReturn] UnicodeEscapeSequence[1,0].IdentifierCodePoint[1200]:RETURN
[NoReturn] UnicodePropertyNameCharacters[0,0].SourceText[2249]:RETURN
[NoReturn] UnicodePropertyNameCharacters[0,1].SourceText[2250]:RETURN
[NoReturn] UnicodePropertyValueCharacters[0,0].SourceText[2251]:RETURN
[NoReturn] UnicodePropertyValueCharacters[0,1].SourceText[2252]:RETURN
[NoReturn] UnicodePropertyValueExpression[0,0].CompileToCharSet[2321]:RETURN
[NoReturn] UnicodePropertyValueExpression[1,0].CompileToCharSet[2322]:RETURN
[NoReturn] UTC[2133]:RETURN
[NoReturn] UTF16SurrogatePairToCodePoint[1192]:RETURN
</details>

<details>
<summary>18 of them contain `new empty List` expression, which evaluates to bottom currently
</summary>

[NoReturn] AsyncBlockStart[2666]:RETURN 
[NoReturn] CaseBlock[1,0].CaseBlockEvaluation[1500]:RETURN
[NoReturn] CaseBlock[1,1].CaseBlockEvaluation[1501]:RETURN
[NoReturn] CyclicModuleRecord.Link[1908]:RETURN
[NoReturn] ExecuteAsyncModule[1914]:RETURN
[NoReturn] ExportDeclaration[2,0].ExportEntries[1982]:RETURN
[NoReturn] ExportDeclaration[3,0].ExportEntries[1983]:RETURN
[NoReturn] ImportDeclaration[1,0].ImportEntries[1936]:RETURN
[NoReturn] MakeArgGetter[1148]:RETURN
[NoReturn] MakeArgSetter[1150]:RETURN
[NoReturn] Module[0,0].ExportEntries[1976]:RETURN
[NoReturn] Module[0,0].ImportEntries[1931]:RETURN
[NoReturn] ModuleItem[0,0].ExportEntries[1978]:RETURN
[NoReturn] ModuleItem[1,0].ImportEntries[1933]:RETURN
[NoReturn] ModuleItem[2,0].ExportEntries[1979]:RETURN
[NoReturn] ModuleItem[2,0].ImportEntries[1934]:RETURN
[NoReturn] NamedExports[0,0].ExportEntriesForModule[1989]:RETURN
[NoReturn] NamedImports[0,0].ImportEntriesForModule[1941]:RETURN
</details>

<details>
<summary>39 of them are not classified automatically, I gave them a manual inspection and classified them manually as much as possible
</summary>

These are related to captured variables in continuations being evaluated to bottom: 

[NoReturn] Await:clo0:cont0[36]:RETURN
[NoReturn] Await:clo1:cont0[38]:RETURN

These are related to access of [[[RemainingElements]] or [[Promise]] internal slots of  Promise related functions:

[NoReturn] INTRINSICS.yet:PromiseRejectFunction[2589]:RETURN
[NoReturn] INTRINSICS.yet:PromiseResolveFunction[2590]:RETURN
[NoReturn] INTRINSICS.yet:`Promise.all`ResolveElementFunction[2605]:RETURN
[NoReturn] INTRINSICS.yet:`Promise.allSettled`RejectElementFunction[2609]:RETURN
[NoReturn] INTRINSICS.yet:`Promise.allSettled`ResolveElementFunction[2608]:RETURN
[NoReturn] INTRINSICS.yet:`Promise.any`RejectElementFunction[2612]:RETURN

These functions appear in the sections for RegExp, which are not currently supported by ESMeta:

[NoReturn] Atom[3,0].CompileAtom[2291]:RETURN 
[NoReturn] Atom[5,0].CompileAtom[2293]:RETURN 
[NoReturn] AtomEscape[0,0].CompileAtom[2294]:RETURN 
[NoReturn] AtomEscape[1,0].CompileAtom[2296]:RETURN 
[NoReturn] CharacterClass[0,0].CompileCharacterClass[2301]:RETURN
[NoReturn] CharacterClass[1,0].CompileCharacterClass[2302]:RETURN
[NoReturn] CharacterClassEscape[4,0].CompileToCharSet[2317]:RETURN
[NoReturn] Term[0,0].CompileSubpattern[2268]:RETURN
[NoReturn] Term[1,0].CompileSubpattern[2269]:RETURN
[NoReturn] RegExpIdentifierName[0,0].CapturingGroupName[2253]:RETURN
[NoReturn] RegExpIdentifierName[1,0].CapturingGroupName[2254]:RETURN
[NoReturn] RegExpIdentifierName[0,0].RegExpIdentifierCodePoints[2255]:RETURN
[NoReturn] RegExpUnicodeEscapeSequence[0,0].CharacterValue[2242]:RETURN
[NoReturn] Pattern[0,0].CompilePattern[2264]:RETURN
[NoReturn] HexLeadSurrogate[0,0].CharacterValue[2245]:RETURN
[NoReturn] HexNonSurrogate[0,0].CharacterValue[2247]:RETURN
[NoReturn] HexTrailSurrogate[0,0].CharacterValue[2246]:RETURN

These are functions that F.Code is Clo that do not contain any AClo:

[NoReturn] BuiltinFunctionObject.Call[1126]:RETURN
[NoReturn] BuiltinFunctionObject.Construct[15]:RETURN

assert (= envRec.SubMap[N] absent) evaluates to False

[NoReturn] DeclarativeEnvironmentRecord.CreateImmutableBinding[1005]:RETURN
[NoReturn] DeclarativeEnvironmentRecord.CreateMutableBinding[1004]:RETURN

These are uncategorized warnings:

[NoReturn] SetDefaultGlobalBindings[1060]:RETURN - (keys @GLOBAL.SubMap) is Bot
[NoReturn] InitializeHostDefinedRealm[1069]:RETURN - caused by SetDefaultGlobalBindings

[NoReturn] Statement[13,0].LabelledEvaluation[1527]:RETURN - https://tc39.es/ecma262/#sec-debugger-statement-runtime-semantics-evaluation
[NoReturn] IdentifierName[0,0].IdentifierCodePoints[1195]:RETURN - caused by YET expression in IdentifierStart[0,0].IdentifierCodePoint
[NoReturn] GeneratorYield[2641]:RETURN - genContext.[[ReturnCont]] is Absent
[NoReturn] IsSharedArrayBuffer[2510]:RETURN - Object.[[ArrayBufferData]] is absent.

[NoReturn] Completion[46]:RETURN - `(? completionRecord: "CompletionRecord")` is bottom.
[NoReturn] CreateResolvingFunctions[2588]:RETURN - caused by `SetFunctionLength` which is already ignored.
[NoReturn] INTRINSICS.Proxy.revocable[2682]:RETURN - caused by `CreateBuiltinFunction` which is already ignored.
</details>

### CODE REVIEW

File: src/main/scala/esmeta/cfg/Func.scala [hunk 1]
```diff
@@ -64,6 +64,10 @@ case class Func(
   lazy val isMethod: Boolean =
     irFunc.kind == FuncKind.ConcMeth || irFunc.kind == FuncKind.InternalMeth
 
+  /** check whether it is closure */
+  lazy val isCloure: Boolean =
```
Comment (hyp3rflow) [inline src/main/scala/esmeta/cfg/Func.scala]:
```suggestion
  lazy val isClosure: Boolean =
```

Comment (doehyunbaek) [reply src/main/scala/esmeta/cfg/Func.scala]: Good catch! fixed in [4468439](https://github.com/es-meta/esmeta/pull/159/commits/4468439781e817c683a385cf7ad2466c5a705bc3)

**Comment by jhnaldo** [comment]
@doehyunbaek, Can you manually rebase your branch into the `dev` branch?

**Comment by doehyunbaek** [comment]
Sure!

**Review by hyp3rflow — APPROVED** [review]

File: src/main/scala/esmeta/lang/util/Parser.scala [hunk 2]
```diff
@@ -1202,6 +1202,7 @@ trait Parsers extends IndentParsers {
 
   // rarely used expressions
   lazy val specialTy: P[Ty] = opt("an " | "a ") ~> {
+    "sequence of Unicode code points" ^^^ StrT |
```
Comment (kimjg1119) [inline src/main/scala/esmeta/lang/util/Parser.scala:1205]: Is the sequence of Unicode code points is String type?

**Review by kimjg1119 — APPROVED** [review]

File: src/main/scala/esmeta/lang/util/Parser.scala (hunk 2 above)
Comment (doehyunbaek) [reply src/main/scala/esmeta/lang/util/Parser.scala:1205]: This is in line with current [tymodel](https://github.com/es-meta/esmeta/blob/9acc5430167456f2c98f473c720340974dfcb5a4/src/main/scala/esmeta/ty/TyModel.scala#L358), which treats EcmascriptFunctionObject.[[SourceText]] as a StrT(https://tc39.es/ecma262/2022/#table-internal-slots-of-ecmascript-function-objects).

**Pull request merged**